### ISSUE TITLE
Inconstent alerts, looking for debug methods

### ISSUE THREAD

**Issue by falz** [nttgin/BGPalerter#946]
We've been doing some targetted testing of the last few days for BGPalerter, specifically for route announcements and withdrawals. This goal is to get a feel of how consistently it will let us know of 'real issues'.

We have reports going to slack, email, syslog. 

Our config.py is auto generated from v1.30 a day or two ago, changes from defaults are:
* uncomment `monitorNewPrefix` in monitors:
* enable syslog, slack, email addresses/webooks (all 'channels' are selected for these)

We announce several dozen prefixes. For testing, we've been announcing and withdrawing a /23 and a /24 within a /16. We then changed prefixes.yml to only monitor that single /16 for further testing, and behaviour is the same - right now it's perhaps one out of every 30/40ish announcements or withdrawals at best. We've done this at misc rates - a few changes in the same hour to waiting until overnight to make a change. This seems to have no bearing on when the alert will or wont come in as it generally simply doesn't.

```yaml
216.56.0.0/16:
  description: testing
  asn:
    - 2381
  ignoreMorespecifics: false
  ignore: false
  group: noc
```

The process is taking about 1.5G of memory.
```
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
bgpaler+  8965  4.5  2.7 1514832 665636 ?      Ssl  09:59   1:35 /opt/bgpalerter/bgpalerter-linux-x64

# pmap 8965 | tail -1
 total          1512132K
```

For debugging purposes, we're monitoring https://ris-live.ripe.net/ using a modified version of the python script on that page (can provide if interested to anyone, it just monitors all data for source ASN). This always shows the changes, and it's approximately 70 updates for each time we make a change, which is above the `monitorVisibility` threshold: `thresholdMinPeers: 40` .

In short, we're looking to see what if any debug flags we can give BGPalerter to help determine where this, and potentially other issues lie.

**Comment by falz** [comment]
Lastly the code we're using to externally monitor the RIS live feed separately from BGPAlerter, just adjust *my_asns* for ones own use (montiors all data with origin ASN)

```python
#!/usr/bin/env python3

#    see RIPE ris-live data filtered by asn
#    python3 -m pip install websocket-client

import json
import websocket
from datetime import datetime

my_asns = [ 2381, 2382 ]

ris_url = "wss://ris-live.ripe.net/v1/ws/?client=py-example-1-modified"

ws = websocket.WebSocket()
ws.connect(ris_url)

params = {
    "moreSpecific": True,
    "host": "rrc21",
    "socketOptions": {
        "includeRaw": True
    }
}

ws.send(json.dumps({ "type": "ris_subscribe", "data": params } ))

for data in ws:
    parsed = json.loads(data)
    if "path" in parsed["data"]:
      if parsed["data"]["path"]:
        source_asn = parsed["data"]["path"].pop()
        if source_asn in my_asns:
          now = datetime.now()
          timestamp = now.strftime("%Y-%m-%d %H:%M:%S")
          data = parsed['data']
          print(timestamp, "ASN:", source_asn, data['type'], "ANNOUNCEMENT:", data['announcements'], "WITHDRAWAL:", data['withdrawals'])
```

**Comment by massimocandela** [comment]
Hi @falz,

By default BGPalerter is set to reduce at minimum the alert sent. It will report the same event about the same prefix only once per `notificationIntervalSeconds` and not every single flap. The default is 24 hours, this configuration targets NOCs, where less is better. For your experiment, set `notificationIntervalSeconds` to something small, like 1 and set all the `thresholdMinPeers` to 1 as well. Then you are basically just filtering RIS. You can also do a step further and set the environment to `research` to disable even more restrictions. Restarting the application is not going to solve anything if you have `persistStatus: true` in your config. RIS beacons are excluded from monitoring (they are used for connection testing in background).

All of this is based on the assumption that your config.yml and prefixes.yml file is configured correctly. No errors in logs/error.log, and the application reports correctly all the prefixes -and ASes- monitored.

**Comment by falz** [comment]
Thank you! I'm now getting consistent test results after lowering `notificationIntervalSeconds` (lowered it to 60) for testing. This will let us properly test further, much appreciated.

I've noticed what may be an actual bug now that I'm testing - the message i receive when we withdraw a prefix is the same as we announce one. Basically it says 'newprefix' for a withdraw OR announcement.

>newprefix
Possible change of configuration. A new prefix 216.56.251.0/24 is announced by AS2381. It is a more specific of 216.56.0.0/16 (testing)

Above was us no longer announcing it.  This may be an issue with the RIS data itsself however as the data (from my sctript) doesnt list it as a withdraw:

`2022-10-25 12:34:33 ASN: 2381 UPDATE ANNOUNCEMENT: [{'next_hop': '37.49.237.194', 'prefixes': ['216.56.251.0/24']}] WITHDRAWAL: []`

And one full update from https://ris-live.ripe.net/ makes it look like they think it's an announcement, not a withdrawal. Perhaps an issue I should take up with RIPE?

```json
// Received at 12:35:10 (4.3 second delay)
{
    "timestamp": 1666719306.12,
    "peer": "195.208.209.200",
    "peer_asn": "199524",
    "id": "195.208.209.200-0184103651880005",
    "host": "rrc13.ripe.net",
    "type": "UPDATE",
    "path": [199524, 31133, 174, 1299, 2381, 2381],
    "community": [],
    "origin": "IGP",
    "announcements": [
        {
            "next_hop": "195.208.209.200",
            "prefixes": [
                "216.56.251.0/24"
            ]
        }
    ],
    "withdrawals": []
}
```

**Comment by massimocandela** [comment]
Hi @falz,

Please, give a look to the history of your BGP updates here: https://stat.ripe.net/data/bgp-updates/data.json?resource=216.56.251.0/24&start_time=2022-10-25&end_time=2022-10-26 I see plenty of withdrawals of your prefixes.
Maybe you are pointing to the wrong update? I struggle to believe RIS is confusing announcements with withdrawals. RIS is used by hundreds of operators and researchers for 20 years now. However, they did a major lifting a few days ago, so if you believe you found an issue, report it at ris@ripe.net

**Comment by massimocandela** [comment]
I'm going to close this. Please, report back if you find an answer.

**Comment by falz** [comment]
Wrote this up (but neglected to submit) just before closed. Commenting anyway.

I do indeed see about ~2k `type": "W"` on that link, but 12k `type": "A"`. I did reach out to RIPE yesterday to ask as well, waiting to hear back. I'm not sure why there would be a discrepancy by 10k updates in that data feed, but perhaps that's normal.

All that I know is if it's supposed to be listed in `"withdrawals": []`, it's not, and the alarm from bgpalerter said it was an advertisement, not a withdrawal.

I tested again while typing this, but with 216.56.252.0/23, and got the same 'newprefix' alert, when the change we made was most definitely removing it.

>newprefix
Possible change of configuration. A new prefix 216.56.252.0/23 is announced by AS2381. It is a more specific of 216.56.0.0/16 (wn testing)

```
[edit routing-options static]
!     inactive: route 216.56.252.0/23 { ... }
```

I'll look a bit more at JSON data and wait to hear back from RIPE. If anyone else following along wants to double check this, please do. I'll also test this with the RIPE beacon prefixes.

Follow up question on your first note - `notificationIntervalSeconds` - where is the 'memory' of alerts held? I see there is a `.cache` directory but prior to me setting that lower, it had nothing in it. I would stop and start bgpalerter but it would still not alarm. Is there a cache of some type stored outside of that area?

**Comment by massimocandela** [comment]
I think I now understand what you are doing and it is not correct.

> I do indeed see about ~2k `type": "W"` on that link, but 12k `type": "A"`.

That is normal. What you see is what the peer is propagating to the RRC (which is the best path to reach your AS).
If the peer is not able to see at all the prefix, you will see a W. Otherwise you will see one or more path changes.

> All that I know is if it's supposed to be listed in `"withdrawals": []`, it's not, and the alarm from bgpalerter said it was an advertisement, not a withdrawal.

What you should monitor is for withdrawals, not to monitor for updates and expect them to be withdrawals. BGPalerter is telling you again you have a newprefix, because you removed all his "memory" by setting a low value of `notificationIntervalSeconds`, so each update, including path changes, are now considered new visibility. At some point, if and when the visibility on at least 3 peers is gone, you will get a visibility alert. You can easily see what I mean by loading your prefix and time range in BGPlay, you will see all the path changes and you will see when the visibility goes down. Check the RIS page for more info on the data collected.

> 
> I tested again while typing this, but with 216.56.252.0/23, and got the same 'newprefix' alert, when the change we made was most definitely removing it.

See my answer above.

> Follow up question on your first note - `notificationIntervalSeconds` - where is the 'memory' of alerts held? I see there is a `.cache` directory but prior to me setting that lower, it had nothing in it. I would stop and start bgpalerter but it would still not alarm. Is there a cache of some type stored outside of that area?

BGPalerter keeps an optimized structure in memory, which gets persisted/restored on exit/start in `.cache`.

**Comment by falz** [comment]
Thanks for details. For whatever reason, I wanst receiving alerts even though nothing was in .cache when i had that timer set longer. I thought I was doing a full stop and start of the process, but perhaps it was just a HUP.

anyhow, setting that notificationIntervalSeconds really is /was my key to testing. Now that I've set that lower, put in my 'real' config that specifies each advertisement as well as parent prefixes, it seems to be working as intended - **visibility** alerts now appear for a specific /23 or /24 as long as I have that prefix listed. 

If I had only the parent /16 prefix listed, I get **announcement** alerts any time it became visible or not, better than nothing.

One additional note that I've noticed with the RIPE beacons, for whcih we're monitoring specific /24's- I only reveive **visibility** alerts (when they are removed) and i never see **announcement** alerts. Must be by design.  

Thanks for the help
